doc/Moving-to-single-app-project-structure.md

```diff
@@ -0,0 +1,160 @@
+# Moving to single application project structure
+
+This document explains how MongooseIM codebase was migrated from the umbrella application structure
+to a single application `rebar3` project.
+
+## Moving official MongooseIM to the new structure
+
+### Move all apps/ejabberd subdirectories to the repo root
+
+- mv `apps/ejabberd/src/* src/`
+- mv `apps/ejabberd/include/* include`
+- mv `apps/ejabberd/priv/* priv`
+- mv `apps/ejabberd/test/* test`
+- mv `apps/ejabberd/c_src/* c_src`
+- mv `apps/ejabberd/asn1/* asn1`
+
+When committing make sure that no new files were added (i.e. they are marked as `renamed`).
+
+### Move tools files
+
+#### cover.spec
+
+`mv apps/ejabberd/src/cover.spec .`
+
+#### rebar.config.script
+
+1. from `apps/ejabberd/src` copy the `MaybeFIPSSupport/1` function
+2. paste it into `rebar.config.script`
+3. call it on a result of calling `SetupIncludedApps/2` in `rebar.config.script`
+
+#### rebar.config
+
+1. In `rebar.config` remove `{i, "apps"}` line from `erl_opts` tuple
+2. Move `{d, xml_nif}` and `{parse_tranform, lager_transform}` lines to `erl_opts` tuple in `rebar.config`
+3. From `apps/ejabberd/rebar.config` move all three `erl_first_files`, `xref_checks` and `port_specs` tuples
+   to `rebar.config`
+4. Add `{cover_print_enabled, true}` line below `{conver_enabled, ...` line in `rebar.config`
+5. In `rebar.config`, from `overrides` tuple remove whole `[{override, ejabberd...` structure
+6. In `rebar.config`, `pre` tuple, add `{compile, {pc, compile}}` tuple (after `{compile, {asn..`)
+7. In `rebar.config`, `post` tuple, add `{clean, {pc, clean}}` tuple (after `{clean, {asn..`)
+
+#### Delete apps/
+
+`rm -rf apps/`
+
+### Changes in source files
+
+#### Application resource files
+
+1. In `src/ejabberd.app.src`
+  * Change application name to `mongooseim`
+  * Change `vsn` to `{cmd, "tools/generate_vsn.sh"}`
+2. Delete `src/mongooseim.app.src.disabled`
+3. `mv src/ejabberd.app.src src/mongooseim.app.src`
+
+### Includes
+
+This is the big one. Since we don't use the umbrella structure anymore, all `-include(..)`
+directives pointing to `ejabberd/include` need to be rewritten.
+
+(SED is `sed` on Linuxes or `gsed` on Macs)
+
+* `gsed -i 's/-include_lib("ejabberd\/include\//-include("/' -- src/*.erl src/*.hrl include/*.hrl test/*.erl`
+* `gsed -i 's/-include_lib("ejabberd\/src\//-include("/' -- src/*.erl src/*.hrl include/*.hrl test/*.erl`
+
+### Release scripts
+
+In `rel/files/mongooseim` the variable `EJABBERD_SO_PATH` contains `ejabberd-2.1.8*` fragment.
+Replace it with `mongooseim-*`.
+
+In `rel/files/mongooseimctl` the variable `EJABBERD_EBIN_PATH` contains `ejabberd-*` fragment.
```

Comment:
In `rel/files/mongooseimctl` the variable `EJABBERD_EBIN_PATH` contains an `ejabberd-*` fragment.